wallet-new/integration/Functions.hs

```diff
@@ -35,454 +37,446 @@ import qualified Pos.Wallet.Web.ClientTypes.Types as V0
 import           Error
 import           Types
 
-
--- | The top function that we need to run in order
--- to test the backend.
+-- | This function uses the given 'ActionProbabilities' to generate a list
+-- of actions to test against the API. It runs the actions through,
+-- eventually returning the final state.
 runActionCheck
     :: (WalletTestMode m, HasCallStack)
     => WalletClient m
     -> WalletState
     -> ActionProbabilities
     -> m WalletState
 runActionCheck walletClient walletState actionProb = do
-    action <- chooseAction actionProb
-    runAction walletClient walletState action
+    actions <- toList <$> chooseActions 20 actionProb
+    let client' = hoistClient lift walletClient
+    execStateT (tryAll (map (runAction client') actions)) walletState
 
+-- | Attempt each action in the list. If an action fails, ignore the
+-- failure.
+tryAll :: Alternative f => [f a] -> f a
+tryAll []     = empty
+tryAll (x:xs) = foldr (\act acc -> act *> acc <|> acc) x xs
 
 -- | Here we run the actions.
 {-# ANN module ("HLint: ignore Reduce duplication" :: Text) #-}
 runAction
-    :: (WalletTestMode m, HasCallStack)
+    :: (WalletTestMode m, HasCallStack, MonadState WalletState m)
     => WalletClient m
-    -> WalletState
     -> Action
-    -> m WalletState
+    -> m ()
 -- Wallets
-runAction wc ws  PostWallet = do
-    newPassword <- liftIO $ generate arbitrary
-    newWall     <- liftIO $ generate $ generateNewWallet newPassword
-    result      <- respToRes $ postWallet wc newWall
+runAction wc action = do
+    log $ "Running: " <> show action
+    case action of
```

Comment:
I switched to an explicit `case` rather than a top level pattern match, because I wanted to log the action that was taken, and also reduce a tiny bit of the boilerplate.